dbt/models/default/default.vw_pin_sale.sql

```diff
@@ -249,93 +244,176 @@ sales_val AS (
         sf.run_id AS sv_run_id,
         sf.version AS sv_version
     FROM
-        {{ source('sale', 'flag') }}
-            AS sf
+        {{ source('sale', 'flag') }} AS sf
     INNER JOIN max_version_flag AS mv
         ON sf.meta_sale_document_num = mv.meta_sale_document_num
         AND sf.version = mv.max_version
+),
+
+combined_sales AS (
+    SELECT
+        COALESCE(uq_sales.pin, md_sales.pin) AS pin,
+        COALESCE(uq_sales.year, md_sales.year) AS year,
+        COALESCE(uq_sales.township_code, tc.township_code) AS township_code,
+        COALESCE(uq_sales.nbhd, tc.nbhd) AS nbhd,
+        COALESCE(uq_sales.class, tc.class) AS class,
+        COALESCE(
+            CASE
+                WHEN md_sales.sale_date < '2021-01-01' THEN md_sales.sale_date
+                WHEN uq_sales.sale_date >= '2021-01-01' THEN uq_sales.sale_date
+            END,
+            md_sales.sale_date,
+            uq_sales.sale_date
+        ) AS sale_date,
```

Comment:
`CASE
WHEN uq_sales.year < '2021' THEN COALESCE(md_sales.sale_date, uq_sales.sale_date)
ELSE COALESCE(uq_sales.sale_date, md_sales.sale_date)
END AS sale_date`

This is how I would do sale date... does that make sense? I'd avoid comparing the sales dates themselves since they might be different and lead to weird behavior

Reply:
We're also gonna have to make sure we use the same logic for `sale_filter_same_sale_within_365`

Reply:
Can you think of a simpler solution than this?

```sql
CASE
    -- If there is a previous sale date within the same partition, we will perform a day difference calculation
    WHEN LAG(
        CASE
            WHEN uq_sales.year < '2021' THEN COALESCE(md_sales.sale_date, uq_sales.sale_date)
            ELSE COALESCE(uq_sales.sale_date, md_sales.sale_date)
        END
    ) OVER (
        -- Define the partition or grouping for the LAG function.
        -- Partition by pin and sale price
        PARTITION BY
            COALESCE(uq_sales.pin, md_sales.pin),
            COALESCE(uq_sales.sale_price, md_sales.sale_price)
        -- Order the sales by the sale date (using the same conditional logic as above) in ascending order
        ORDER BY
            CASE
                WHEN uq_sales.year < '2021' THEN COALESCE(md_sales.sale_date, uq_sales.sale_date)
                ELSE COALESCE(uq_sales.sale_date, md_sales.sale_date)
            END ASC
    ) IS NOT NULL
    -- If there is a previous sale, calculate difference in days between the current sale date and the previous sale date.
    THEN
        -- Use EXTRACT to compute the day difference between the current sale date and the previous sale date.
        EXTRACT(
            DAY FROM 
                -- The current sale date
                CASE
                    WHEN uq_sales.year < '2021' THEN COALESCE(md_sales.sale_date, uq_sales.sale_date)
                    ELSE COALESCE(uq_sales.sale_date, md_sales.sale_date)
                END
            -- Subtract the previous sale date, which is found using the LAG function.
            - LAG(
                CASE
                    WHEN uq_sales.year < '2021' THEN COALESCE(md_sales.sale_date, uq_sales.sale_date)
                    ELSE COALESCE(uq_sales.sale_date, md_sales.sale_date)
                END
            ) OVER (
                PARTITION BY
                    COALESCE(uq_sales.pin, md_sales.pin),
                    COALESCE(uq_sales.sale_price, md_sales.sale_price)
                ORDER BY
                    CASE
                        WHEN uq_sales.year < '2021' THEN COALESCE(md_sales.sale_date, uq_sales.sale_date)
                        ELSE COALESCE(uq_sales.sale_date, md_sales.sale_date)
                    END ASC
            )
        ) <= 365
    -- If there is no previous sale in the same partition (first sale or no prior sale at the same price for that pin), return FALSE.
    ELSE FALSE
END AS sale_filter_same_sale_within_365
```

Reply:
You could construct it outside of the `combined_sales` CTE, in the main `SELECT` statement, just using `combined_sales.sale_date` etc.